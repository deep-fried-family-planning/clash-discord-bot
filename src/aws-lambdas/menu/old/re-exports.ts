export {
    ApplicationCommandType as CMD,
    ApplicationCommandOptionType as CMD_OP,
    InteractionType as ITR,
    MessageFlags as MSG,
    WebhookType as WBH,
    TextInputStyle as CMP_T,
    // ButtonStyle as CMP_B,
} from '@discordjs/core/http-only';

import {Discord} from 'dfx';

export type CMP = Discord.ComponentType;
export const CMP = Discord.ComponentType;
export type CMP_B = Discord.ButtonStyle;
export const CMP_B = Discord.ButtonStyle;
