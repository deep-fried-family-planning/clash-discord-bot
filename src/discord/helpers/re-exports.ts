export {
    ApplicationCommandType as CMD,
    ApplicationCommandOptionType as CMD_OP,
    InteractionType as ITR,
    MessageFlags as MSG,
    WebhookType as WBH,
    ComponentType as CMP,
    TextInputStyle as CMP_T,
    ButtonStyle as CMP_B,
} from 'discord-api-types/v10';
