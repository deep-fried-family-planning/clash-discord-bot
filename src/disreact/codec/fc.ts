import * as Const from '#src/disreact/model/internal/core/enum.ts';
import * as Proto from '#src/disreact/model/internal/infrastructure/proto.ts';
import * as E from 'effect/Effect';
import * as Equal from 'effect/Equal';
import * as Hash from 'effect/Hash';
import * as Predicate from 'effect/Predicate';

export const TypeId  = Symbol('disreact/fc'),
             SYNC    = 1,
             PROMISE = 2,
             EFFECT  = 3,
             KindId  = Symbol('disreact/fc/kind'),
             NameId  = Symbol('disreact/fc/name');

export namespace FC {
  type Thing<A, B> = (props: A) => B;


  export interface Any<P, O> extends Thing<P, O> {
    (props: P): O | Promise<O> | E.Effect<O, any, any>;
    [TypeId]?   : FC;
    [KindId]?   : number;
    [NameId]?   : string;
    displayName?: string;
  }
  export interface Sync<P, O> extends Any<P, O> {
    (props: P): O;
    [KindId]: typeof SYNC;
    [NameId]: string;
  }
  export interface Async<P, O> extends Any<P, O> {
    (props: P): O | Promise<O>;
    [KindId]: typeof PROMISE;
    [NameId]: string;
  }
  export interface Effect<P, O> extends Any<P, O> {
    (props: P): E.Effect<O, any, any>;
    [KindId]: typeof EFFECT;
    [NameId]: string;
  }
  export type FC<P = any, O = any> =
    | FC.Any<P, O>
    | FC.Sync<P, O>
    | FC.Async<P, O>
    | FC.Effect<P, O>;
}
export type Any<P = any, O = any> = FC.Any<P, O>;
export type Sync<P = any, O = any> = FC.Sync<P, O>;
export type Async<P = any, O = any> = FC.Async<P, O>;
export type Effect<P = any, O = any> = FC.Effect<P, O>;
export type FC<P = any, O = any> = FC.FC<P, O>;

export const isFC = (fc: unknown): fc is FC => typeof fc === 'function';

export const isDefined = (fc: unknown): fc is FC =>
  typeof fc === 'function'
  && TypeId in fc
  && fc[TypeId] === TypeId;

const FcProto = {
  [Hash.symbol](this: FC) {
    return Hash.string(TypeId.toString());
  },
  [Equal.symbol](this: FC, that: FC) {
    return this[TypeId] === that[TypeId];
  },
};

export const type = (fc: FC) => fc[TypeId]!;
const kind = (fc: FC) => fc[KindId];

export const make = (f: Any): Any => {
  if (f[TypeId]) {
    return f;
  }
  if (!f[NameId]) {
    if (f.displayName) {
      f[NameId] = f.displayName;
    }
    else if (f.name) {
      f[NameId] = f.name;
    }
    else {
      f[NameId] = '.';
    }
    if (f.constructor.name === Const.ASYNC_FUNCTION) {
      f[KindId] = Const.PROMISE;
    }
  }
  (f[TypeId] as any) = TypeId;
  return f;
};

export const id = (input: FC.FC) => input[NameId];

export const name = (fc: FC.FC, name: string) => {
  const current = id(fc);
  if (current === name) {
    return;
  }
  if (current === '.') {
    fc[NameId] = name;
  }
  else {
    throw new Error(`Renaming ${current} to ${name}`);
  }
};

export const render = (f: FC, p: any): E.Effect<any> => {
  switch (f[KindId]) {
    case Const.SYNC: {
      return E.sync(() => f(p));
    }
    case Const.PROMISE: {
      return E.promise(() => f(p));
    }
    case Const.EFFECT: {
      return f(p);
    }
    default: {
      return E.suspend(() => {
        const out = f(p);
        if (Predicate.isPromise(out)) {
          f[KindId] = Const.PROMISE;
          return E.promise(() => out);
        }
        if (E.isEffect(out)) {
          f[KindId] = Const.EFFECT;
          return out as E.Effect<any>;
        }
        f[KindId] = Const.SYNC;
        return E.succeed(out);
      });
    }
  }
};
