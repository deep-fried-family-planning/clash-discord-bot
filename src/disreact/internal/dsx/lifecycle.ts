/* eslint-disable @typescript-eslint/no-explicit-any,@typescript-eslint/no-unsafe-argument,@typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-member-access */
import {onautocomplete, onclick, ondeselect, oninvoke, onselect, onsubmit} from '#src/disreact/internal/dsx/datt.ts';
import {dsxid} from '#src/disreact/internal/dsx/dsx.ts';
import {__dismount, __get, __mount, __prep} from '#src/disreact/internal/dsx/globals.ts';
import type {T} from '#src/disreact/abstract/event.ts';
import {DTML} from '#src/disreact/internal/dsx/index.ts';
import type {Hooks, HooksById, HookStacksById} from '#src/disreact/internal/types.ts';
import {Data, Equal} from 'effect';
import console from 'node:console';



export type RenderFn = (props: any) => any;

type Common = {
  index       : number;
  name        : string;
  id          : string;
  id_step     : string;
  id_full     : string;
  isRoot?     : boolean;
  isModal?    : boolean;
  isMessage?  : boolean;
  isEphemeral?: boolean;
};

export type PragmaText = Common & {
  kind : 'text';
  name : 'string';
  value: string;
};

export type PragmaElement = Common & {
  kind    : 'intrinsic';
  name    : keyof JSX.IntrinsicElements;
  props   : Record<string, any>;
  children: Pragma[];
};

export type PragmaFunction = Common & {
  kind    : 'function';
  props   : Record<string, any>;
  children: Pragma[];
  render  : RenderFn;
  state?  : Hooks;
  stack?  : Hooks['stack'];
};

export type Pragma =
  | PragmaText
  | PragmaElement
  | PragmaFunction;



export const cloneTree = (node: Pragma, parent?: Pragma): Pragma => {
  const baseClone = dsxid(node, parent);

  if (node.kind === 'text') {
    return structuredClone(baseClone);
  }

  if (node.kind === 'intrinsic') {
    const {children: _, props, ...rest} = baseClone as PragmaElement;

    const clone = structuredClone(rest);

    const {[onclick]: __, [onselect]: ___, [ondeselect]: ____, [onsubmit]: ______, [oninvoke]: _______, [onautocomplete]: ________, ...restProps} = props;

    const clonedProps = structuredClone(restProps);

    for (const fn of [onclick, onselect, ondeselect, onsubmit, oninvoke, onautocomplete]) {
      if (props[fn]) {
        clonedProps[fn] = props[fn];
      }
    }

    return {
      ...clone,
      props   : clonedProps,
      children: node.children.map((child) => cloneTree(child, {...node, ...baseClone})),
    } as PragmaElement;
  }

  const {children: _, render, ...rest} = baseClone as PragmaFunction;
  const clone                          = structuredClone(rest);

  return {
    ...clone,
    children: node.children.map((child) => cloneTree(child, {...node, ...baseClone})),
    render,
  } as PragmaFunction;
};



export const initialRender = (node: Pragma, parent?: Pragma): Pragma => {
  const base = dsxid(node, parent);

  switch (node.kind) {
    case 'text': {
      return base;
    }
    case 'intrinsic': {
      return {
        ...base,
        children: node.children.map((child) => initialRender(child, base)),
      } as PragmaElement;
    }
    case 'function': {
      const children = renderFunctionNode(node);

      return {
        ...base,
        kind    : 'function',
        children: children.map((child: any) => initialRender(child, base)),
        render  : node.render,
        state   : node.state,
      } as PragmaFunction;
    }
  }
};



export const dispatchEvent = (node: Pragma, event: T, original: Pragma = node): Pragma => {
  console.log('[dispatchEvent]', node.id_step);
  if ('props' in node) {
    if (node.id_step === event.id && node.props[event.type]) {
      if (typeof node.props[event.type] === 'function') {
        node.props[event.type](event);
        return original;
      }
    }
  }

  if ('children' in node) {
    for (const child of node.children) {
      try {
        return dispatchEvent(child, event, original);
      }
      catch (_) {
        // do nothing
      }
    }
  }

  throw new Error(`No node with id_step "${event.id}" having a handler for type "${event.type}" was not found`);
};



export const hydrateRoot = (node: Pragma, states: {[k: string]: Hooks}): Pragma => {
  if (node.kind === 'text') {
    return node;
  }
  if (node.kind === 'intrinsic') {
    node.children = setIds(node.children, node).map((child) => hydrateRoot(child, states));
    return node;
  }

  if (node.id_full in states) {
    node.state = states[node.id_full];
    node.stack = structuredClone(node.state.stack);
  }

  node.children = renderFunctionNode(node).map((child) => hydrateRoot(child, states));
  return node;
};



export const rerenderRoot = (root: Pragma): Pragma => {
  if (root.kind === 'text') {
    return root;
  }
  if (root.kind === 'intrinsic') {
    return root;
  }
  const current  = root.children;
  const rendered = renderFunctionNode(root);
  root.children  = renderNodes(root, current, rendered);
  return root;
};

const renderNodes = (parent: Pragma, cs: Pragma[], rs: Pragma[]): Pragma[] => {
  const length   = Math.max(cs.length, rs.length);
  const children = [];

  for (let i = 0; i < length; i++) {
    const c = cs.at(i);
    const r = rs.at(i);

    if (!c && !r) {
      throw new Error('Impossible');
    }
    if (!c && r) {
      __mount(r.id_full);
      children.push(initialRender(r, parent));
      continue;
    }
    if (c && !r) {
      __dismount(c.id_full);
      continue;
    }
    if (c && r) {
      if (!isSameNode(c, r)) {
        __dismount(c.id_full);
        children.push(initialRender(r, parent));
      }
      else if (c.kind === 'text') {
        children.push(c);
      }
      else if (c.kind === 'intrinsic') {
        if (!hasSameProps(c, r)) {
          c.props = (r as PragmaElement).props;
        }
        c.children = renderNodes(c, c.children, (r as PragmaElement).children);
        children.push(c);
      }
      else if (hasSameProps(c, r)) {
        if (hasSameState(c)) {
          c.children = renderNodes(c, c.children, c.children);
          children.push(c);
        }
        else {
          const nextchildren = renderFunctionNode(c);
          c.children         = renderNodes(c, c.children, nextchildren);
          children.push(c);
        }
      }
      else {
        const nextchildren = renderFunctionNode(c);
        c.children         = renderNodes(c, c.children, nextchildren);
        children.push(c);
      }
    }
  }
  return children;
};

const renderFunctionNode = (node: PragmaFunction) => {
  __prep(node.id_full, node.state);

  const output = node.render(node.props) as any[];

  const children = Array.isArray(output) ? output : [output] as any[];
  const rendered = setIds(children, node);
  node.state     = __get(node.id_full);
  node.stack     = structuredClone(node.state.stack);
  node.state.pc  = 0;
  node.state.rc++;

  if (children.some((child) => child.name === DTML.dialog || child.name === DTML.modal)) {
    node.isModal = true;
  }
  if (children.some((child) => child.name === DTML.message)) {
    node.isMessage = true;
    if (children.some((child) => child.props.ephemeral)) {
      node.isEphemeral = true;
    }
  }

  return rendered;
};



const isSameNode = <A extends Pragma, B extends Pragma>(a: A, b: B) => {
  console.debug('[isSameNode]', a.id_step, b.id_step);
  if (a.kind !== b.kind) return false;
  if (a.name !== b.name) return false;
  if (a.id !== b.id) return false;
  if (a.kind === 'text') return a.value === (b as PragmaText).value;
  console.debug('[isSameNode]: true');
  return true;
};

const hasSameProps = (c: Pragma, r: Pragma) => {
  const cprops = Data.struct((c as PragmaFunction | PragmaElement).props);
  const rprops = Data.struct((r as PragmaFunction | PragmaElement).props);
  const equals = Equal.equals(cprops, rprops);
  console.debug('[hasSameProps]', cprops, rprops, equals);
  return equals;
};

const hasSameState = (c: Pragma) => {
  const connected = Data.array((c as PragmaFunction).state!.stack.map(Data.struct));
  const last      = Data.array((c as PragmaFunction).stack!.map(Data.struct));
  const equals    = Equal.equals(connected, last);
  console.debug('[hasSameState]', connected, last, equals);
  return equals;
};



const setIds = (children: Pragma[], parent: Pragma) => {
  for (let i = 0; i < children.length; i++) {
    children[i].index = i;
    children[i]       = dsxid(children[i], parent);
  }
  return children;
};



export const collectStates = (node: Pragma, states: HooksById = {}): HooksById => {
  if (node.kind === 'function' && node.state) {
    states[node.state.id] = node.state;
  }

  if ('children' in node && Array.isArray(node.children)) {
    for (const child of node.children) {
      collectStates(child, states);
    }
  }

  return states;
};

export const reduceToStacks = (hooks: Record<string, Hooks>): HookStacksById => {
  return Object.fromEntries(
    Object.entries(hooks)
      .filter(([_, value]) => value.stack.length)
      .map(([key, value]) => [key, value.stack]),
  );
};
