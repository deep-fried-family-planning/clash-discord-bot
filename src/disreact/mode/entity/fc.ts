/* eslint-disable @typescript-eslint/no-unsafe-function-type */
import type * as El from '#src/disreact/mode/entity/el.ts';
import * as E from 'effect/Effect';
import * as P from 'effect/Predicate';

export const TypeId = Symbol.for('disreact/fc/type');
export const NameId = Symbol.for('disreact/fc/name');
export const SYNC = 0;
export const PROMISE = 1;
export const EFFECT = 2;

export namespace FC {
  export type Output = El.Cs;
  export type OutEffect = E.Effect<El.Cs>;

  export interface Any<P, O> extends Function {
    (props: P): O | Promise<O> | E.Effect<O, any, any>;
    [TypeId]?   : number;
    [NameId]?   : string;
    displayName?: string;
  }
  export interface Sync<P, O> extends Any<P, O> {
    (props: P): O;
    [TypeId]: typeof SYNC;
    [NameId]: string;
  }
  export interface Async<P, O> extends Any<P, O> {
    (props: P): O | Promise<O>;
    [TypeId]: typeof PROMISE;
    [NameId]: string;
  }
  export interface Effect<P, O> extends Any<P, O> {
    (props: P): E.Effect<O, any, any>;
    [TypeId]: typeof EFFECT;
    [NameId]: string;
  }
  export type FC<P = any, O = FC.Output> =
    | FC.Any<P, O>
    | FC.Sync<P, O>
    | FC.Async<P, O>
    | FC.Effect<P, O>;
}
export type Any<P = any, O = FC.Output> = FC.Any<P, O>;
export type Sync<P = any, O = FC.Output> = FC.Sync<P, O>;
export type Async<P = any, O = FC.Output> = FC.Async<P, O>;
export type Effect<P = any, O = FC.Output> = FC.Effect<P, O>;
export type FC<P = any, O = FC.Output> = FC.FC<P, O>;

export const make = (fc: FC.Any<any, any>) => {
  if (fc[TypeId] || fc[NameId]) return fc;
  if (fc.displayName) fc[NameId] = fc.displayName;
  else if (fc.name) fc[NameId] = fc.name;
  else fc[NameId] = '.';
  if (fc.constructor.name === 'AsyncFunction') fc[TypeId] = PROMISE;
  return fc;
};

export const isFC = (fc: unknown): fc is FC => typeof fc === 'function';

export const name = (input: FC.FC) => input[NameId]!;

export const render = (fc: FC.FC, props: any = {}) => {
  if (fc[TypeId] === SYNC) {
    return E.sync(() => fc(props)) as FC.OutEffect;
  }
  if (fc[TypeId] === PROMISE) {
    return E.promise(async () => await fc(props)) as FC.OutEffect;
  }
  if (fc[TypeId] === EFFECT) {
    return fc(props) as FC.OutEffect;
  }
  return E.suspend(() => {
    const out = fc(props);
    if (P.isPromise(out)) {
      fc[TypeId] = PROMISE;
      return E.promise(async () => await out) as FC.OutEffect;
    }
    if (E.isEffect(out)) {
      fc[TypeId] = EFFECT;
      return out as FC.OutEffect;
    }
    fc[TypeId] = SYNC;
    return E.succeed(out) as FC.OutEffect;
  });
};
