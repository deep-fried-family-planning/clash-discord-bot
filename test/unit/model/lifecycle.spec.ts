import {Trigger} from '#src/disreact/model/elem/trigger.ts';
import {Lifecycle} from '#src/disreact/model/lifecycle.ts';
import {Registry} from '#src/disreact/model/Registry.ts';
import {Rehydrant} from '#src/disreact/model/rehydrant.ts';
import {flow, S} from '#src/disreact/utils/re-exports.ts';
import {E} from '#src/internal/pure/effect.ts';
import {MessageEffect} from '#test/unit/components/message-effect.tsx';
import {MessageAsync} from '#test/unit/components/message-async.tsx';
import {MessageSync} from '#test/unit/components/message-sync.tsx';
import {Record} from 'effect';
import {TestMessage} from '#test/unit/components/test-message.tsx';
import {it} from '#test/unit/components/TestRegistry.tsx';
import { FC } from '#src/disreact/model/meta/fc.ts';

const json = (input: any) => JSON.stringify(input, null, 2);
const snap = (root: Rehydrant) => json(root);
const toStacks = (root: Rehydrant) => Record.map(root.fibrils, (v) => v.stack);
const hash = flow(Rehydrant.dehydrate, S.encodeSync(Rehydrant.Hydrator));

it.effect('when rendering sync', E.fn(function* () {
  const root = yield* Registry.checkout(MessageSync);
  yield* Lifecycle.initialize(root);
  const encoding = yield* Lifecycle.encode(root);
  expect(snap(encoding?.data)).toMatchSnapshot(FC.getName(MessageSync));
}));

it.effect('when rendering async', E.fn(function* () {
  const root = yield* Registry.checkout(MessageAsync);
  yield* Lifecycle.initialize(root);
  const encoding = yield* Lifecycle.encode(root);
  expect(snap(encoding?.data)).toMatchSnapshot(FC.getName(MessageAsync));
}));

it.effect('when rendering effect', E.fn(function* () {
  const root = yield* Registry.checkout(MessageEffect);
  yield* Lifecycle.initialize(root);
  const encoding = yield* Lifecycle.encode(root);
  expect(snap(encoding?.data)).toMatchSnapshot(FC.getName(MessageEffect));
}));

it.effect('when initial rendering', E.fn(function* () {
  const root = yield* Registry.checkout(TestMessage);
  yield* Lifecycle.initialize(root);
  const encoding = yield* Lifecycle.encode(root);
  expect(snap(encoding?.data)).toMatchSnapshot();
}));

it.effect('when dispatching an event', E.fn(function* () {
  const registry = yield* Registry;
  const root = yield* registry.checkout(TestMessage);

  yield* Lifecycle.initialize(root);

  expect(toStacks(root)).toMatchSnapshot('initial stacks');
  expect(hash(root)).toMatchSnapshot('initial hash');
  const initial = yield* Lifecycle.encode(root);
  expect(snap(initial?.data)).toMatchSnapshot('initial encoded');

  const event = Trigger.make('actions:2:button:0', {});
  yield* Lifecycle.invoke(root, event);
  yield* Lifecycle.rerender(root);

  expect(toStacks(root)).toMatchSnapshot('rerendered stacks');
  expect(hash(root)).toMatchSnapshot('rerendered hash');
  const rerendered = yield* Lifecycle.encode(root);
  expect(snap(rerendered?.data)).toMatchSnapshot('rerendered encoded');
}));

describe('given event.id does not match any node.id', () => {
  it.effect('when dispatching an event', E.fn(function* () {
    const registry = yield* Registry;
    const root = yield* registry.checkout(TestMessage);
    yield* Lifecycle.initialize(root);
    yield* Lifecycle.rerender(root);

    const event = Trigger.make('buttons:1:button:0', {});

    expect(() => E.runSync(Lifecycle.invoke(root, event) as any)).toThrowErrorMatchingSnapshot();
  }));
});

it.effect(`when hydrating an empty root (performance)`, E.fn(function* () {
  const runs = Array.from({length: 10000});

  for (let i = 0; i < runs.length; i++) {
    const root = yield* Registry.checkout(TestMessage);
    yield* Lifecycle.initialize(root);
    yield* Lifecycle.rehydrate(root);

    const event = Trigger.make('actions:2:button:0', {});

    yield* Lifecycle.invoke(root, event);
    yield* Lifecycle.rerender(root);

    yield* Lifecycle.encode(root);
  }
}), {timeout: 10000});
