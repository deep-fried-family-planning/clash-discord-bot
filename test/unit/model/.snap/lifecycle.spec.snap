// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`given event.id does not match any node.id > when dispatching an event 1`] = `[(FiberFailure) EventDefect: An error has occurred]`;

exports[`when dispatching an event > initial encoded 1`] = `
"{
  "embeds": [
    {
      "title": "Omni Board",
      "description": "V2 - JSX Pragma"
    },
    {
      "title": "Test Title - NOT markdown",
      "description": "### H3-# \\n<@123456>\\n__$ not a link__"
    }
  ],
  "components": [
    {
      "type": 1,
      "components": [
        {
          "type": 2,
          "style": 1,
          "custom_id": "actions:0:button:0",
          "label": "Start 0"
        },
        {
          "type": 2,
          "style": 2,
          "custom_id": "actions:0:secondary:0",
          "label": "Help"
        }
      ]
    },
    {
      "type": 1,
      "components": [
        {
          "type": 3,
          "custom_id": "message:0:select:0",
          "options": [
            {
              "value": "1",
              "label": "1"
            },
            {
              "value": "2",
              "label": "2"
            },
            {
              "value": "3",
              "label": "3",
              "default": true
            }
          ]
        }
      ]
    }
  ],
  "flags": 64
}"
`;

exports[`when dispatching an event > initial hash 1`] = `"eJxrXpSZsjoktbjEN7W4ODE9dWlBUX5BccOy4pLE5OziprVIUlYGExsXFjPcVEARs8qFszxSE1NSi6wMJgAAI4UifQ"`;

exports[`when dispatching an event > initial stacks 1`] = `
{
  "id": "TestMessage",
  "props": {
    Symbol(disreact/props): Symbol(disreact/props),
    Symbol(effect/Hash): [Function],
    Symbol(effect/Equal): [Function],
  },
  "stacks": {
    "TestMessage:0": [
      {
        "s": 0,
      },
    ],
    "TestMessage:0:message:0:Header:0": [],
  },
}
`;

exports[`when dispatching an event > rerendered encoded 1`] = `
"{
  "embeds": [
    {
      "title": "Omni Board",
      "description": "V2 - JSX Pragma"
    },
    {
      "title": "Test Title - NOT markdown",
      "description": "### H3-# \\n<@123456>\\n__$ not a link__"
    }
  ],
  "components": [
    {
      "type": 1,
      "components": [
        {
          "type": 2,
          "style": 1,
          "custom_id": "actions:0:button:0",
          "label": "Start 1"
        },
        {
          "type": 2,
          "style": 2,
          "custom_id": "actions:0:secondary:0",
          "label": "Help"
        }
      ]
    },
    {
      "type": 1,
      "components": [
        {
          "type": 3,
          "custom_id": "message:0:select:0",
          "options": [
            {
              "value": "1",
              "label": "1"
            },
            {
              "value": "2",
              "label": "2"
            },
            {
              "value": "3",
              "label": "3",
              "default": true
            }
          ]
        }
      ]
    }
  ],
  "flags": 64
}"
`;

exports[`when dispatching an event > rerendered hash 1`] = `"eJxrXpSZsjoktbjEN7W4ODE9dWlBUX5BccOy4pLE5OziprVIUlYGExsXFjPeVEARs8qFszxSE1NSi6wMJgAAI6kifg"`;

exports[`when dispatching an event > rerendered stacks 1`] = `
{
  "id": "TestMessage",
  "props": {
    Symbol(disreact/props): Symbol(disreact/props),
    Symbol(effect/Hash): [Function],
    Symbol(effect/Equal): [Function],
  },
  "stacks": {
    "TestMessage:0": [
      {
        "s": 1,
      },
    ],
    "TestMessage:0:message:0:Header:0": [],
  },
}
`;

exports[`when initial rendering 1`] = `
"{
  "embeds": [
    {
      "title": "Omni Board",
      "description": "V2 - JSX Pragma"
    },
    {
      "title": "Test Title - NOT markdown",
      "description": "### H3-# \\n<@123456>\\n__$ not a link__"
    }
  ],
  "components": [
    {
      "type": 1,
      "components": [
        {
          "type": 2,
          "style": 1,
          "custom_id": "actions:0:button:0",
          "label": "Start 0"
        },
        {
          "type": 2,
          "style": 2,
          "custom_id": "actions:0:secondary:0",
          "label": "Help"
        }
      ]
    },
    {
      "type": 1,
      "components": [
        {
          "type": 3,
          "custom_id": "message:0:select:0",
          "options": [
            {
              "value": "1",
              "label": "1"
            },
            {
              "value": "2",
              "label": "2"
            },
            {
              "value": "3",
              "label": "3",
              "default": true
            }
          ]
        }
      ]
    }
  ],
  "flags": 64
}"
`;

exports[`when rendering async > MessageAsync 1`] = `
"{
  "embeds": [
    {
      "description": "MessageAsync"
    }
  ],
  "components": [
    {
      "type": 1,
      "components": [
        {
          "type": 2,
          "style": 1,
          "custom_id": "ButtonSync:0:button:0",
          "label": "ButtonSync"
        },
        {
          "type": 2,
          "style": 1,
          "custom_id": "ButtonAsync:0:button:0",
          "label": "ButtonAsync"
        },
        {
          "type": 2,
          "style": 1,
          "custom_id": "ButtonEffect:0:button:0",
          "label": "ButtonEffect"
        }
      ]
    }
  ],
  "flags": 64
}"
`;

exports[`when rendering effect > MessageEffect 1`] = `
"{
  "embeds": [
    {
      "description": "EphemeralEffect"
    }
  ],
  "components": [
    {
      "type": 1,
      "components": [
        {
          "type": 2,
          "style": 1,
          "custom_id": "ButtonSync:0:button:0",
          "label": "ButtonSync"
        },
        {
          "type": 2,
          "style": 1,
          "custom_id": "ButtonAsync:0:button:0",
          "label": "ButtonAsync"
        },
        {
          "type": 2,
          "style": 1,
          "custom_id": "ButtonEffect:0:button:0",
          "label": "ButtonEffect"
        }
      ]
    }
  ],
  "flags": 64
}"
`;

exports[`when rendering sync > MessageSync 1`] = `
"{
  "embeds": [
    {
      "description": "MessageSync"
    }
  ],
  "components": [
    {
      "type": 1,
      "components": [
        {
          "type": 2,
          "style": 1,
          "custom_id": "ButtonSync:0:button:0",
          "label": "ButtonSync"
        },
        {
          "type": 2,
          "style": 1,
          "custom_id": "ButtonAsync:0:button:0",
          "label": "ButtonAsync"
        },
        {
          "type": 2,
          "style": 1,
          "custom_id": "ButtonEffect:0:button:0",
          "label": "ButtonEffect"
        }
      ]
    }
  ]
}"
`;

exports[`when rendering sync 1`] = `
{
  "$d": 0,
  "$i": 0,
  "$p": 0,
  "_n": "MessageSync:0",
  "_s": "MessageSync:0",
  "name": "MessageSync",
  "props": {
    Symbol(disreact/props): Symbol(disreact/props),
    Symbol(effect/Hash): [Function],
    Symbol(effect/Equal): [Function],
  },
  "rs": [
    {
      "$d": 1,
      "$i": 0,
      "$p": 0,
      "_n": "MessageSync:0:message:0",
      "_s": "MessageSync:0:message:0",
      "handler": undefined,
      "name": "message",
      "props": {
        Symbol(disreact/props): Symbol(disreact/props),
        Symbol(effect/Hash): [Function],
        Symbol(effect/Equal): [Function],
      },
      "rs": [
        {
          "$d": 2,
          "$i": 0,
          "$p": 0,
          "_n": "MessageSync:0:message:0:embed:0",
          "_s": "message:0:embed:0",
          "handler": undefined,
          "name": "embed",
          "props": {
            Symbol(disreact/props): Symbol(disreact/props),
            Symbol(effect/Hash): [Function],
            Symbol(effect/Equal): [Function],
          },
          "rs": [
            {
              "$d": 3,
              "$i": 0,
              "$p": 0,
              "_n": "MessageSync:0:message:0:embed:0:undefined:0",
              "_s": "embed:0:undefined:0",
              "text": "MessageSync",
              Symbol(disreact/element): 1,
              Symbol(effect/Hash): [Function],
              Symbol(effect/Equal): [Function],
            },
          ],
          "type": "embed",
          Symbol(disreact/element): 2,
          Symbol(effect/Hash): [Function],
          Symbol(effect/Equal): [Function],
        },
        {
          "$d": 2,
          "$i": 0,
          "$p": 1,
          "_n": "MessageSync:0:message:0:actions:0",
          "_s": "message:0:actions:0",
          "handler": undefined,
          "name": "actions",
          "props": {
            Symbol(disreact/props): Symbol(disreact/props),
            Symbol(effect/Hash): [Function],
            Symbol(effect/Equal): [Function],
          },
          "rs": [
            {
              "$d": 3,
              "$i": 0,
              "$p": 0,
              "_n": "MessageSync:0:message:0:actions:0:ButtonSync:0",
              "_s": "actions:0:ButtonSync:0",
              "name": "ButtonSync",
              "props": {
                Symbol(disreact/props): Symbol(disreact/props),
                Symbol(effect/Hash): [Function],
                Symbol(effect/Equal): [Function],
              },
              "rs": [
                {
                  "$d": 4,
                  "$i": 0,
                  "$p": 0,
                  "_n": "MessageSync:0:message:0:actions:0:ButtonSync:0:button:0",
                  "_s": "ButtonSync:0:button:0",
                  "handler": [Function],
                  "name": "button",
                  "props": {
                    "label": "ButtonSync",
                    Symbol(disreact/props): Symbol(disreact/props),
                    Symbol(effect/Hash): [Function],
                    Symbol(effect/Equal): [Function],
                  },
                  "type": "button",
                  Symbol(disreact/element): 2,
                  Symbol(effect/Hash): [Function],
                  Symbol(effect/Equal): [Function],
                },
              ],
              "type": [Function],
              Symbol(disreact/element): 3,
              Symbol(effect/Hash): [Function],
              Symbol(effect/Equal): [Function],
            },
            {
              "$d": 3,
              "$i": 0,
              "$p": 1,
              "_n": "MessageSync:0:message:0:actions:0:ButtonAsync:0",
              "_s": "actions:0:ButtonAsync:0",
              "name": "ButtonAsync",
              "props": {
                Symbol(disreact/props): Symbol(disreact/props),
                Symbol(effect/Hash): [Function],
                Symbol(effect/Equal): [Function],
              },
              "rs": [
                {
                  "$d": 4,
                  "$i": 0,
                  "$p": 0,
                  "_n": "MessageSync:0:message:0:actions:0:ButtonAsync:0:button:0",
                  "_s": "ButtonAsync:0:button:0",
                  "handler": [Function],
                  "name": "button",
                  "props": {
                    "label": "ButtonAsync",
                    Symbol(disreact/props): Symbol(disreact/props),
                    Symbol(effect/Hash): [Function],
                    Symbol(effect/Equal): [Function],
                  },
                  "type": "button",
                  Symbol(disreact/element): 2,
                  Symbol(effect/Hash): [Function],
                  Symbol(effect/Equal): [Function],
                },
              ],
              "type": [Function],
              Symbol(disreact/element): 3,
              Symbol(effect/Hash): [Function],
              Symbol(effect/Equal): [Function],
            },
            {
              "$d": 3,
              "$i": 0,
              "$p": 2,
              "_n": "MessageSync:0:message:0:actions:0:ButtonEffect:0",
              "_s": "actions:0:ButtonEffect:0",
              "name": "ButtonEffect",
              "props": {
                Symbol(disreact/props): Symbol(disreact/props),
                Symbol(effect/Hash): [Function],
                Symbol(effect/Equal): [Function],
              },
              "rs": [
                {
                  "$d": 4,
                  "$i": 0,
                  "$p": 0,
                  "_n": "MessageSync:0:message:0:actions:0:ButtonEffect:0:button:0",
                  "_s": "ButtonEffect:0:button:0",
                  "handler": [Function],
                  "name": "button",
                  "props": {
                    "label": "ButtonEffect",
                    Symbol(disreact/props): Symbol(disreact/props),
                    Symbol(effect/Hash): [Function],
                    Symbol(effect/Equal): [Function],
                  },
                  "type": "button",
                  Symbol(disreact/element): 2,
                  Symbol(effect/Hash): [Function],
                  Symbol(effect/Equal): [Function],
                },
              ],
              "type": [Function],
              Symbol(disreact/element): 3,
              Symbol(effect/Hash): [Function],
              Symbol(effect/Equal): [Function],
            },
          ],
          "type": "actions",
          Symbol(disreact/element): 2,
          Symbol(effect/Hash): [Function],
          Symbol(effect/Equal): [Function],
        },
      ],
      "type": "message",
      Symbol(disreact/element): 2,
      Symbol(effect/Hash): [Function],
      Symbol(effect/Equal): [Function],
    },
  ],
  "type": [Function],
  Symbol(disreact/element): 3,
  Symbol(effect/Hash): [Function],
  Symbol(effect/Equal): [Function],
}
`;
